aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java

```diff
@@ -0,0 +1,430 @@
+import com.amplifyframework.core.Consumer;
+
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.mobile.client.AWSMobileClient;
+import com.amazonaws.mobile.client.Callback;
+import com.amazonaws.mobile.client.HostedUIOptions;
+import com.amazonaws.mobile.client.IdentityProvider;
+import com.amazonaws.mobile.client.SignInUIOptions;
+import com.amazonaws.mobile.client.UserState;
+import com.amazonaws.mobile.client.UserStateDetails;
+import com.amazonaws.mobile.client.results.SignInResult;
+import com.amazonaws.mobile.client.results.SignInState;
+import com.amazonaws.mobile.client.results.SignUpResult;
+import com.amazonaws.mobile.client.results.Tokens;
+import com.amazonaws.mobile.client.results.UserCodeDeliveryDetails;
+import com.amazonaws.mobile.config.AWSConfiguration;
+import org.json.JSONObject;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public final class AWSCognitoAuthPlugin extends AuthPlugin<AWSMobileClient> {
+    private static final String AWS_COGNITO_AUTH_PLUGIN_KEY = "awsCognitoAuthPlugin";
+    private static final long SECONDS_BEFORE_TIMEOUT = 10;
+
+    @NonNull
+    @Override
+    public String getPluginKey() {
+        return AWS_COGNITO_AUTH_PLUGIN_KEY;
+    }
+
+    @Override
+    public void configure(
+            @NonNull JSONObject pluginConfiguration,
+            @NonNull Context context
+    ) throws AmplifyException {
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        AWSMobileClient.getInstance().initialize(
+            context,
+            new AWSConfiguration(pluginConfiguration),
+            new Callback<UserStateDetails>() {
+                @Override
+                public void onResult(UserStateDetails result) {
+                    latch.countDown();
+                }
+
+                @Override
+                public void onError(Exception error) {
+                    throw new IllegalStateException("Failed to instantiate AWSMobileClient", error);
+                }
+            }
+        );
+
+        try {
+            if (latch.await(SECONDS_BEFORE_TIMEOUT, TimeUnit.SECONDS)) {
+                return;
+            } else {
+                throw new IllegalStateException(
+                    "Failed to instantiate AWSMobileClient within " + SECONDS_BEFORE_TIMEOUT + " seconds"
+                );
+            }
+        } catch (InterruptedException error) {
+            throw new IllegalStateException(error);
+        }
+    }
+
+    public void signUp(
+            String username,
+            String password,
+            AuthSignUpOptions options,
+            final Consumer<AuthSignUpResult> onSuccess,
+            final Consumer<AuthException> onException
+    ) {
+        AWSMobileClient.getInstance().signUp(
+            username,
+            password,
+            options.getUserAttributes(),
+            options.getValidationData(),
+            new Callback<SignUpResult>() {
```

Comment:
Hmm, I see what you're saying but I think especially since this is only a temporary implementation of the plugin wrapping AWSMobileClient for GA and the plan is to do a fast follow with a completely fresh new Auth implementation, it probably isn't worth doing now.

Reply:
And while there are a couple extra lines of code added this way vs. the way you recommend, it's actually faster to write because the IDE autocompletes it.